site-builder/server/server.ts

```diff
@@ -0,0 +1,586 @@
+ * @param s3Prefix - S3 key prefix (folder path)
+ * @param timestamp - Timestamp to use in the prefix (for versioning)
+ * @returns Base URL of the uploaded content
+ */
+const uploadDirectoryToS3 = async (
+	sourceDir: string,
+	s3Prefix: string,
+	timestamp?: number
+): Promise<string> => {
+	const bucket = env.S3_BUCKET_NAME;
+
+	// Use timestamp for versioning if provided
+	const prefix = timestamp ? `${s3Prefix}/${timestamp}` : s3Prefix;
+
+	console.log(`Starting directory upload to S3: ${sourceDir} → s3://${bucket}/${prefix}`);
+
+	// Get list of all files recursively
+	const getAllFiles = async (dir: string): Promise<{ path: string; relativePath: string }[]> => {
+		const entries = await fs.readdir(dir, { withFileTypes: true });
+		const files = await Promise.all(
+			entries.map(async (entry) => {
+				const fullPath = path.join(dir, entry.name);
+				const relativePath = path.relative(sourceDir, fullPath);
+
+				if (entry.isDirectory()) {
+					return getAllFiles(fullPath);
+				} else {
+					return [
+						{
+							path: fullPath,
+							relativePath,
+						},
+					];
+				}
+			})
+		);
+
+		return files.flat();
+	};
+
+	// Get all files with their paths
+	const allFiles = await getAllFiles(sourceDir);
+	const totalFiles = allFiles.length;
+
+	console.log(`Found ${totalFiles} files to upload`);
+
+	// Upload files in parallel with max concurrency
+	const concurrencyLimit = 5;
+	const results = [];
+	let completedFiles = 0;
+
+	// Process files in batches for controlled concurrency
+	for (let i = 0; i < allFiles.length; i += concurrencyLimit) {
+		const batch = allFiles.slice(i, i + concurrencyLimit);
+		const batchPromises = batch.map(async (file) => {
+			try {
+				const contentType = mime.lookup(file.path) || "application/octet-stream";
+
+				// Convert Windows paths to forward slashes for S3
+				const s3Key = `${prefix}/${file.relativePath.replace(/\\/g, "/")}`;
+
+				// Create a read stream instead of reading the whole file into memory
+				const fileStream = createReadStream(file.path);
+
+				// Upload the file stream
+				await uploadFileToS3(s3Prefix, file.relativePath.replace(/\\/g, "/"), fileStream, {
+					contentType: contentType,
+				});
+
+				completedFiles++;
+				if (completedFiles % 10 === 0 || completedFiles === totalFiles) {
+					console.log(
+						`Uploaded ${completedFiles}/${totalFiles} files (${Math.round((completedFiles / totalFiles) * 100)}%)`
+					);
+				}
+
+				return s3Key;
+			} catch (error) {
+				console.error(`Error uploading ${file.path}:`, error);
+				throw error;
+			}
+		});
+
+		// Wait for current batch to complete before starting next batch
+		const batchResults = await Promise.all(batchPromises);
+		results.push(...batchResults);
+	}
+
+	console.log(
+		`Directory upload complete. Uploaded ${results.length} files to s3://${bucket}/${prefix}`
+	);
+
+	// Return the base URL of the uploaded content
+	return `${env.S3_ENDPOINT}/${bucket}/${prefix}`;
+};
```

Comment:
the idea was: what if you could just upload all the files to `bucket/prefix` and now you have a site! `minio` dont work like that, and our s3 also doesnt, so maybe remove this in the future